### PR TITLE
Check Actvity

### PR DESCRIPTION
**Pull request by arsalaerfan**
## Check Activity Endpoint

This PR implements an activity checking endpoint for Accelerate students that monitors both **attendance** and **Canvas engagement** to determine student activity status. It automatically creates `accelerate_course_progress` records for students who don't have one, or updates existing records with the latest Canvas access timestamp.

### Issues Fixed
* https://github.com/NickGuerrero/cti-sys/issues/66

### How It Works

The endpoint checks students who are active in the main student roster (`students.active = True`) and enrolled in Accelerate, then updates their `accelerate.active` status based on recent activity.

- Recent attendance at Accelerate sessions *(configurable threshold in weeks)*
- Recent Canvas login access *(configurable threshold in weeks)*

**Activity Logic**

Students are marked as **active** in `accelerate.active` if they have **either**:
- Attended an Accelerate session within the threshold period, **OR**
- Accessed Canvas within the threshold period

### Database Changes

**Migration:** Added `last_canvas_access` field to `accelerate_course_progress` table:

**Apply Migration:**

1. Make sure DB URL has `postgresql+psycopg`:
```bash
   export DEV_DATABASE_URL="postgresql+psycopg://.../devdb"
```

2. Run the migration:
```bash
   alembic -x environment=development upgrade head
```

### Tests
`TestCheckAccelerateActivity` pytest includes 9 total tests:

- All activity combinations (both, attendance only, Canvas only, none)
- Edge cases (no students, missing Canvas ID)
- Error handling (Canvas API failures, graceful per-student rollback)
- Database operations (create/update accelerate_course_progress records)

**To run tests locally:**
```
pytest -v -k "TestCheckAccelerateActivity"
```

Screenshot 2025-12-11 at 4 39 23 PM

**On Github Actions:**
```
pytest -v -m "not integration"
```
Screenshot 2025-12-11 at 4 41 42 PM

### Not Implemented (Future Enhancement)

**Program Breaks / Skip Weeks:** This implementation does not currently account for program breaks where students aren't required to attend sessions, as initially requested in the issue. We will revisit this enhancement at a later time.

### Checklist
- [x] I've reviewed the contribution guide
- [x] I've confirmed the changes work on my machine
- [x] This pull request is ready for review

### CODE REVIEW

File: migrations/env.py [hunk 1]
```diff
@@ -29,7 +29,7 @@
 
 def get_url() -> str:
     # Alembic Command line: alembic -x environment=development
-    env_name = config.get_main_option("environment") or "development"
+    env_name = context.get_x_argument(as_dictionary=True).get("environment") or "development"
```
Comment (NickGuerrero) [inline migrations/env.py:32]: I made this change since the included instructions were wrong; it would default to the alembic.ini file instead of the command line argument. It does highlight a important decision regarding the environment variables in issue #49.

File: src/config.py [hunk 2]
```diff
@@ -46,6 +46,14 @@ class Settings(BaseSettings):
...
     gs_project_id: Optional[str] = Field(validation_alias="GS_PROJECT_ID", default=None)
 
+    # Thresholds in weeks for determining if a student is active
+    activity_attendance_threshold_weeks: int = Field(validation_alias="ACTIVITY_ATTENDANCE_THRESHOLD_WEEKS",
+                                                     default=2,
+                                                     description="Number of weeks to check for attendance activity")
+    activity_canvas_threshold_weeks: int = Field(validation_alias="ACTIVITY_CANVAS_THRESHOLD_WEEKS",
+                                                 default=2,
+                                                 description="Number of weeks to check for Canvas activity")
+
```
Comment (NickGuerrero) [inline src/config.py:56]: I adjusted this to fit in line with the other env variables in the config.py. Although redundant, I prefer explicitly calling the environment variables, and leaving it as a simple Field can unintentionally fetch a value and override the threshold. While I had intended for this to be a simple constant, enabling environment override should be fine.

File: src/students/accelerate/check_activity/service.py [hunk 3]
```diff
@@ -0,0 +1,210 @@
...
+        timeout=10,
+    )
+    
+    if response.status_code == 404:
+        return None
+    
+    if response.status_code == 401:
+        raise ValueError(
+            "Please check that CTI_ACCESS_TOKEN is set correctly and has not expired."
+        )
```
Comment (NickGuerrero) [inline src/students/accelerate/check_activity/service.py:42]: Note that specific to Canvas, there's a special response code, ```429 Forbidden (Rate Limit Exceeded)```. ([More throttling details here](https://developerdocs.instructure.com/services/canvas/basics/file.throttling)). This was part of the reason I asked for the rate limiting solution as the following issue.

This solution should be sufficient for now, since our student roster currently fits under the rate limit comfortably. I wonder if there was a way to get multiple in a single call, but that may not have been possible with the API. Regardless, we'll revisit this in the future.

File: src/students/accelerate/check_activity/service.py [hunk 4]
```diff
@@ -0,0 +1,210 @@
...
+                else:
+                    results["students_marked_inactive"] += 1
+                
+                results["details"].append(result)
+                
+        except Exception as e:
+            db.rollback()
+            results["errors"].append({"cti_id": student.cti_id, "error": str(e)})
+    
+    return results
```
Comment (NickGuerrero) [inline src/students/accelerate/check_activity/service.py:210]: In the future, ideally results are sent to a logger and the response can be a summary of operations. This currently works since we don't have a logger, but it would be helpful if we could adjust the logic later to avoid having large responses to sift through.

Comment (NickGuerrero) [reply src/students/accelerate/check_activity/service.py:210]: I will also add that this isn't as much of a problem for production, since the response is never sent, however the result object is still created and in memory, which isn't necessary if it's never used.

**Review by NickGuerrero — APPROVED** [review]
* Endpoint route and service looks clean and straightforward
* Migration done correctly, small issue resolved
* Provided tests pass, however there's a lack of integration tests

Everything works in development, but taking this into the production server makes me a bit concerned. We do have database back-ups in case anything goes wrong. Although not specifically requested, integration tests help alleviate this fear, although they're difficult to set-up (I might be the only one with access to everything).

Everything looks good and is clear to merge into main. With this PR and the improvement in #70, we should be ready for the next version release.
